### PR TITLE
Fix #3268: Introduce Maven pinning artifacts Mechanism

### LINKED ISSUE #3268: Introduce Pinning Arifacts Mechanism for Maven dependencies
Integrate Maven artifacts pinning to generate maven_install.json, the new source of truth for the Maven dependencies used in the repo.

**Target PR date**: 2021-06-09
**Target completion date**: 2021-06-13

### PR DESCRIPTION
**Pull request by prayutsu**
## Explanation

Fixes #3268:
Introduced Pinning Artifacts mechanism for Maven dependencies.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by prayutsu** [comment]
@rt4914 PTAL.

File: WORKSPACE
Comment (prayutsu) [inline WORKSPACE:150]: When I tried setting `fail_if_repin_required` attribute here - 
```
maven_install(
    artifacts = DAGGER_ARTIFACTS + get_maven_dependencies(),
    fetch_sources = True,
    maven_install_json = "//:maven_install.json",
    fail_if_repin_required = True,
    repositories = DAGGER_REPOSITORIES + [
        "https://maven.fabric.io/public",
        "https://maven.google.com",
        "https://repo1.maven.org/maven2",
    ],
)
```

It threw an error - 
```
Starting local Bazel server and connecting to it...
ERROR: Traceback (most recent call last):
        File "/home/prayutsu/opensource/oppia-android/WORKSPACE", line 145, column 14, in <toplevel>
                maven_install(
        File "/home/prayutsu/.cache/bazel/_bazel_prayutsu/65c630ff6a7321024a3fd83ac1880b0b/external/rules_jvm_external/defs.bzl", line 20, column 5, in maven_install
                def maven_install(
Error: maven_install() got unexpected keyword argument: fail_if_repin_required
ERROR: error loading package '': Encountered error while reading extension file 'defs.bzl': no such package '@maven//': error loading package 'external': Could not load //external package

```

Screenshot from 2021-06-07 18-38-33

**Comment by prayutsu** [comment]
It looks like the `fail_if_repin_reqired` attribute isn't set properly (as Bazel Lint test is failing and also Build is not failing upon editing Maven dependencies).
It seems `fail_if_repin_required` is declared as a variable rather than an attribute.

File: WORKSPACE
Comment (fsharpasharp) [reply WORKSPACE:150]: Try a newer version of rules_jvm_external.

https://github.com/oppia/oppia-android/blob/f87e5047ef9da68fac4bc40168149ff50e1d4a09/third_party/versions.bzl#L112

Comment (prayutsu) [reply WORKSPACE:150]: Got fixed, Thanks @fsharpasharp :) 
We were using quite an old version.

**Comment by prayutsu** [comment]
@rt4914 @BenHenning Should I update the CODEOWNERS file to mention appropriate CODEOWNERS for maven_install.json?

**Comment by prayutsu** [comment]
@rt4914 @fsharpasharp PTAL.

**Review by fsharpasharp — APPROVED** [review]
Looks good to me.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@prayutsu Update the code-owner file to mention Ben's name for maven file. Also assigning this to Ben.

**Comment by seanlip** [comment]
@prayutsu I'm not sure why you're assigning this to me? Ben can approve for codeowners.

**Comment by prayutsu** [comment]
@seanlip Actually I didn't manually add you as a reviewer, maybe you got added as a reviewer because this PR involves changes in the CODEOWNERS file.

**Comment by seanlip** [comment]
Ah ok, thanks for explaining. Makes sense! 

For future reference, you can use the "shield" icon next to the reviewer name to figure out whether the person is a mandatory reviewer or not (in deciding whether to assign them).

**Comment by prayutsu** [comment]
> Ah ok, thanks for explaining. Makes sense!
> 
> For future reference, you can use the "shield" icon next to the reviewer name to figure out whether the person is a mandatory reviewer or not (in deciding whether to assign them).

Noted, thanks Sean.

**Comment by prayutsu** [comment]
Closing and reopening to trigger CI checks.

**Review by rt4914 — APPROVED** [review]
LGTM. Wait for @BenHenning approval.

File: .github/CODEOWNERS
Comment (BenHenning) [inline .github/CODEOWNERS]: Per my feedback in your document today, could we maybe put this in third_party, or do we not have a choice?

File: third_party/versions.bzl
Comment (BenHenning) [inline third_party/versions.bzl:114]: Can you provide some context on why this upgrade is needed?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @prayutsu. Leaving some initial comments.

File: third_party/versions.bzl
Comment (prayutsu) [reply third_party/versions.bzl:114]: The older version didn't support the `fail_if_repin_required` attribute. It has been introduced lately, that's why we need this upgrade.

File: .github/CODEOWNERS
Comment (prayutsu) [reply .github/CODEOWNERS]: I tried moving it to third_party package by following the [documentation](https://github.com/bazelbuild/rules_jvm_external#custom-location-for-maven_installjson) but it is throwing error - `Could not find a BUILD file`

Screenshot from 2021-06-22 11-52-32

I also tried making the same location as mentioned in the documentation (by adding name attribute and making a package in the third-party directory along with an empty BUILD file) but it threw the same error.
Screenshot from 2021-06-22 12-01-13

@fsharpasharp Can you spot some mistakes here? (I have pushed the changes that according to me should do the work here).

File: WORKSPACE
Comment (BenHenning) [inline WORKSPACE]: Shouldn't this be ``//third_party:maven_install.json`` since the file is part of our repository?

File: third_party/versions.bzl
Comment (BenHenning) [reply third_party/versions.bzl:114]: Got it--thanks for the clarification!

File: .github/CODEOWNERS
Comment (BenHenning) [reply .github/CODEOWNERS]: Just left another comment--I think you might have made a mistake in the WORKSPACE file.

**Review by BenHenning — COMMENTED** [review]
@prayutsu apologies for the delayed review. PTAL at my latest comments--I think you might've made a mistake when changing WORKSPACE to point to third_party/maven_install.json.

Also, just to check: doesn't maven_install.json need to be moved to third_party/?

File: WORKSPACE
Comment (prayutsu) [reply WORKSPACE]: I am still getting the same error.
Screenshot from 2021-06-28 21-47-39

File: .github/CODEOWNERS
Comment (prayutsu) [reply .github/CODEOWNERS:75]: Even after changing this, I am getting the same error (I already tried this earlier).

File: WORKSPACE
Comment (prayutsu) [reply WORKSPACE]: Done. Thanks for pointing this out.

File: .github/CODEOWNERS
Comment (prayutsu) [reply .github/CODEOWNERS]: Thanks, it got fixed. I also had to shift it to third_party manually.

**Review by BenHenning — COMMENTED** [review]
@prayutsu I think you need to add a CODEOWNERS entry for third_party/ (that can go to me). Also, it looks like the maven_install.json file needs to be regenerated.

Beyond that, the Gradle test failures seem to be #2844 which is disappointing--I thought that was addressed. Rerunning them should hopefully fix the issue.

**Comment by prayutsu** [comment]
> @prayutsu I think you need to add a CODEOWNERS entry for third_party/ (that can go to me). Also, it looks like the maven_install.json file needs to be regenerated.
> 
> Beyond that, the Gradle test failures seem to be #2844 which is disappointing--I thought that was addressed. Rerunning them should hopefully fix the issue.

Done.

File: .github/CODEOWNERS
Comment (BenHenning) [inline .github/CODEOWNERS]:
```suggestion
# Third-party dependencies.
```

File: third_party/maven_install.json
Comment (BenHenning) [inline third_party/maven_install.json:2995]: Do you know why these are needed now? Is it because of a merge with develop?

**Review by BenHenning — COMMENTED** [review]
Thanks @prayutsu. Just two last quick comments.

File: third_party/maven_install.json
Comment (prayutsu) [reply third_party/maven_install.json:2995]: Yes, right, I took a pull from develop and this dependency was added.

File: .github/CODEOWNERS
Comment (prayutsu) [reply .github/CODEOWNERS]: Done.

**Comment by BenHenning** [comment]
Per previous runs, the compute affected tests script doesn't seem to produce any tests to run here, so skipping it since the last two runs have hung.

**Review by BenHenning — APPROVED** [review]
Thanks @prayutsu. This LGTM!

**Pull request merged**
